### PR TITLE
Infer tuple size from is_tuple/1 and add hint

### PR DESCRIPTION
**Pull request by ericmj**
Closes https://github.com/elixir-lang/elixir/issues/10823.

### CODE REVIEW

File: lib/elixir/test/elixir/module/types/types_test.exs [hunk 1]
```diff
@@ -321,6 +321,33 @@ defmodule Module.Types.TypesTest do
...
+
+                 # types_test.ex:1
+                 is_tuple(foo)
+
+             where "foo" was given the type {dynamic()} in:
+
+                 # types_test.ex:1
+                 {_} = foo
+
+             HINT: use "is_tuple(foo) and tuple_size(foo) == 1" to infer a sized tuple.
```
Comment (josevalim) [inline lib/elixir/test/elixir/module/types/types_test.exs:347]:
```suggestion
             HINT: use "is_tuple(foo) and tuple_size(foo) == 1" to guard a sized tuple.
```

**Review by josevalim — APPROVED** [review]
Beautiful, just one comment :)

File: lib/elixir/lib/module/types.ex [hunk 2]
```diff
@@ -416,18 +415,24 @@ defmodule Module.Types do
     integer() unless said otherwise. For example, <<expr>> assumes "expr" \
     is an integer. Pass a modifier, such as <<expr::float>> or <<expr::binary>>, \
     to change the default behaviour.
+    """
+  end
 
+  defp format_message_hint({:sized_and_unsize_tuples, {size, var}}) do
+    """
+    HINT: use "is_tuple(#{Macro.to_string(var)}) and \
+    tuple_size(#{Macro.to_string(var)}) == #{size}" to infer a sized tuple.
```
Comment (ericmj) [inline lib/elixir/lib/module/types.ex:424]:
```suggestion
    tuple_size(#{Macro.to_string(var)}) == #{size}" to guard a sized tuple.
```

**Comment by eksperimental** [comment]
Awesome! Thank you @ericmj

**Pull request merged**